### PR TITLE
add eventemitter3

### PR DESCRIPTION
**Pull request by luu-alex**
replacing nodejs and web3js implemetation of eventemitter for the usage of eventemitter3 for react-native compatibiltiy. 
## Description

Please include a summary of the changes and be sure to follow our [Contribution Guidelines](https://github.com/web3/web3.js/blob/4.x/.github/CONTRIBUTING.md).

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run lint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.
- [ ] I have linked Issue(s) with this PR in "Linked Issues" menu.

### CODE REVIEW

File: packages/web3-providers-ipc/test/unit/check_implementation.test.ts [hunk 1]
```diff
@@ -97,23 +97,6 @@ describe('IPCProvider', () => {
 			expect(end).toHaveBeenCalled();
 		});
 
-		it('connection error', async () => {
```
Comment (luu-alex) [inline packages/web3-providers-ipc/test/unit/check_implementation.test.ts:100]: removed connection error test, mock implementation doesn't seem to emit error for this one unit test so i added an integration test to substitute it
https://github.com/web3/web3.js/pull/6823/files#diff-ea6da20dfb39c75b95f9a7cb4cf0e3f9d1e353fe0aab69f28ef83497e86ff556R140

File: packages/web3-eth-contract/test/unit/contract.test.ts [hunk 2]
```diff
@@ -755,7 +756,7 @@ describe('Contract', () => {
 
 			const clonnedContract = contract.clone();
 
-			expect(JSON.stringify(contract)).toStrictEqual(JSON.stringify(clonnedContract));
+			expect(stringify(contract)).toStrictEqual(stringify(clonnedContract));
```
Comment (luu-alex) [inline packages/web3-eth-contract/test/unit/contract.test.ts:759]: eventemitter3 has a circular structure so we use stringify library to parse properly

File: packages/web3-providers-ws/test/unit/check_implementation.test.ts [hunk 3]
```diff
@@ -85,28 +85,14 @@ describe('WebSocketProvider', () => {
 			ws.disconnect(code, data);
 			expect(close).toHaveBeenCalledWith(code, data);
 		});
-		it('connection error', async () => {
```
Comment (luu-alex) [inline packages/web3-providers-ws/test/unit/check_implementation.test.ts:88]: removed connection error test, mock implementation doesn't seem to emit error for this one unit test so i added an integration test to substitute it
https://github.com/web3/web3.js/pull/6823/files#diff-ea6da20dfb39c75b95f9a7cb4cf0e3f9d1e353fe0aab69f28ef83497e86ff556R140

File: packages/web3-providers-ws/test/unit/__mocks__/isomorphic-ws.ts [hunk 4]
```diff
@@ -23,10 +23,9 @@ export default class WebSocket extends EventEmitter {
 	public CONNECTING = 0;
 	public OPEN = 1;
 
-	public constructor(...args: any[]) {
+	public constructor() {
```
Comment (jdevcs) [inline packages/web3-providers-ws/test/unit/__mocks__/isomorphic-ws.ts:26]: this will be breaking change

**Review by jdevcs — APPROVED** [review]

File: packages/web3-eth-contract/test/unit/contract.test.ts (hunk 2 above)
Comment (jdevcs) [reply packages/web3-eth-contract/test/unit/contract.test.ts:759]: whats error for that?

**Comment by jdevcs** [comment]
> # Bundle Stats
> Hey there, this message comes from a github action that helps you and reviewers to understand how these changes affect the size of this project's bundle.
> 
> As this PR is updated, I'll keep you updated on how the bundle size is impacted.
> 
> **Total**
> 
> Asset	Old size	New size	Diff	Diff %
> Total	0	615 KB	615 KB	-
> View detailed bundle breakdown

Its not showing difference

**Review by avkos — APPROVED** [review]

**Review by avkos — CHANGES_REQUESTED** [review]
please take a look at tests.

**Comment by luu-alex** [comment]
@avkos  testcases passing, benchmarks is failing on main

**Comment by luu-alex** [comment]
@jdevcs 
I'm not exactly sure but the eventemitter3 build is smaller

main branch min.js build:
web3: $ npx webpack
web3: asset web3.min.js 602 KiB [emitted] [minimized] [big] (name: main) 2 related assets
web3: asset ../lib/commonjs/index.d.ts 8.63 KiB [compared for emit]
web3: asset ../lib/commonjs/accounts.d.ts 3.89 KiB [compared for emit]
web3: asset ../lib/commonjs/types.d.ts 2.45 KiB [compared for emit]
web3: asset ../lib/commonjs/web3.d.ts 1.14 KiB [compared for emit]
web3: asset ../lib/commonjs/abi.d.ts 999 bytes [compared for emit]
web3: asset ../lib/commonjs/eth.exports.d.ts 280 bytes [compared for emit]
web3: asset ../lib/commonjs/providers.exports.d.ts 148 bytes [compared for emit]
web3: asset ../lib/commonjs/version.d.ts 60 bytes [compared for emit]
web3: runtime modules 1010 bytes 5 modules
web3: orphan modules 788 bytes [orphan] 2 modules
web3: modules by path ../ 1.6 MiB 288 modules
web3: modules by path ./src/*.ts 13.7 KiB
web3:   ./src/web3.ts 7.78 KiB [built] [code generated]
web3:   ./src/abi.ts 1.26 KiB [built] [code generated]
web3:   ./src/accounts.ts 4.47 KiB [built] [code generated]
web3:   ./src/version.ts 186 bytes [built] [code generated]
web3: WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
web3: This can impact web performance.
web3: Assets: 
web3:   web3.min.js (602 KiB)
web3: WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
web3: Entrypoints:
web3:   main (602 KiB)
web3:       web3.min.js
web3: 

eventemitter3 min.js build:

> web3:"build:web"

web3: $ npx webpack
web3: asset web3.min.js 599 KiB [emitted] [minimized] [big] (name: main) 2 related assets
web3: asset ../lib/commonjs/index.d.ts 8.63 KiB [compared for emit]
web3: asset ../lib/commonjs/accounts.d.ts 3.89 KiB [compared for emit]
web3: asset ../lib/commonjs/types.d.ts 2.45 KiB [compared for emit]
web3: asset ../lib/commonjs/web3.d.ts 1.26 KiB [compared for emit]
web3: asset ../lib/commonjs/abi.d.ts 999 bytes [compared for emit]
web3: asset ../lib/commonjs/web3_eip6963.d.ts 922 bytes [compared for emit]
web3: asset ../lib/commonjs/eth.exports.d.ts 280 bytes [compared for emit]
web3: asset ../lib/commonjs/providers.exports.d.ts 148 bytes [compared for emit]
web3: asset ../lib/commonjs/version.d.ts 60 bytes [compared for emit]
web3: runtime modules 1010 bytes 5 modules
web3: orphan modules 788 bytes [orphan] 2 modules
web3: modules by path ../ 1.59 MiB 288 modules
web3: modules by path ./src/*.ts 15.5 KiB
web3:   ./src/web3.ts 7.91 KiB [built] [code generated]
web3:   ./src/abi.ts 1.26 KiB [built] [code generated]
web3:   ./src/accounts.ts 4.47 KiB [built] [code generated]
web3:   ./src/version.ts 186 bytes [built] [code generated]
web3:   ./src/web3_eip6963.ts 1.65 KiB [built] [code generated]
web3: WARNING in asset size limit: The following asset(s) exceed the recommended size limit (244 KiB).
web3: This can impact web performance.
web3: Assets: 
web3:   web3.min.js (599 KiB)
web3: WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
web3: Entrypoints:
web3:   main (599 KiB)
web3:       web3.min.js
web3: 
web3: WARNING in webpack performance recommendations: 
web3: You can limit the size of your bundles by using import() or require.ensure to lazy load some parts of your application.
web3: For more info visit https://webpack.js.org/guides/code-splitting/
web3: webpack 5.76.0 compiled with 3 warnings in 11041 ms
tting/
web3: webpack 5.76.0 compiled with 3 warnings in 8970 ms
with latest pull

**Review by avkos — APPROVED** [review]

File: packages/web3-eth-contract/test/unit/contract.test.ts (hunk 2 above)
Comment (luu-alex) [reply packages/web3-eth-contract/test/unit/contract.test.ts:759]:
```
    TypeError: Converting circular structure to JSON
        --> starting at object with constructor 'EventEmitter'
```

**Comment by jdevcs** [comment]
> > # Bundle Stats
> > Hey there, this message comes from a github action that helps you and reviewers to understand how these changes affect the size of this project's bundle.
> > As this PR is updated, I'll keep you updated on how the bundle size is impacted.
> > **Total**
> > Asset	Old size	New size	Diff	Diff %
> > Total	0	615 KB	615 KB	-
> > View detailed bundle breakdown
> 
> Its not showing difference

@avkos and I talked about this lib, and it will increase our lib size.

**Comment by luu-alex** [comment]
@jdevcs  i believe so, the lib size is 73.4 kB so it will not be a drastic change (web3js is 3.27 MB)

**Pull request merged**